### PR TITLE
8303624: The java.lang.Thread.FieldHolder can be null for JNI attaching threads

### PR DESCRIPTION
**Pull request by dholmes-ora**
To support virtual threads a number of fields were moved out of `java.lang.Thread` into a separate `FieldHolder` object. The VM was updated to then access certain thread fields via the `FieldHolder`. 

The code for attaching a thread to the VM, specifically `allocate_threadObj` didn't allow for the `Thread` constructor throwing an exception, and so failing to allocate the `FieldHolder` before attempting to access a field through the `FieldHolder`. This resulted in assertion failures in `javaClasses.cpp` (or crashes in a product build). That code is fixed to ensure we cease processing if the constructor throws an exception.

In addition, we need to recognise that whilst a native thread is attaching via JNI, it is partially initialized (to varying degrees) but also visible through JVMTI. Though the window is small JVMTI could get hold of an attaching thread and then invoke methods that would try to access uninitialized state. When this was state of `Thread` instance there was no problem as the object existed in its zero-initialized form (it had been allocated directly but no constructor run). However, anythng accessed via the `FieldHolder` is now a problem as the `FieldHolder` may not exist. So we modify all of the `FieldHolder` get/set methods to account for it being null: setters will do nothing, while getters return the default zero value for the field.

Testing: tiers 1-3 as a sanity test

There's no way to write a regression test for this.

Thanks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303624](https://bugs.openjdk.org/browse/JDK-8303624): The java.lang.Thread.FieldHolder can be null for JNI attaching threads

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [67246a74](https://git.openjdk.org/jdk/pull/12892/files/67246a74ca6c8eabad5f8be1118011d16407b206)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [67246a74](https://git.openjdk.org/jdk/pull/12892/files/67246a74ca6c8eabad5f8be1118011d16407b206)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12892/head:pull/12892` \
`$ git checkout pull/12892`

Update a local copy of the PR: \
`$ git checkout pull/12892` \
`$ git pull https://git.openjdk.org/jdk pull/12892/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12892`

View PR using the GUI difftool: \
`$ git pr show -t 12892`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12892.diff">https://git.openjdk.org/jdk/pull/12892.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
> There's no way to write a regression test for this.

Clarification: there's no way to write a general regression test for accessing the fieldHolder fields via JVMTI while the thread is attaching. The original failure scenario of the Thread constructor throwing an exception could potentially be testable.

**Comment by AlanBateman** [comment]
If I read this correctly, a JNI attaching thread has its JavaThread on the threads list, the j.l.Thread object is allocated but not yet initialized or the initialization fails (in both cases, holder is null). If a JVMTI agent gets a reference to the thread at this point (GetAllThreads, instrumented the constructor, several other ways) then functions such as GetThreadInfo will attempt to access fields and crash.

The changes look okay to me except for SET_FIELDHOLDER_FIELD being a no-op when holder is null. It should be okay for a JVMTI agent to read a default value but if set_priority,  set_daemon, or set_thread_status can be being called when the initialization fails then it suggests there the error handling might isn't complete somewhere.

**Comment by dholmes-ora** [comment]
@AlanBateman thanks for looking at this. Note that there are no API's for setting the daemon status once the thread is alive (which it is) nor directly changing the thread's status (that can only be done via actions of the thread itself). That only leaves an API call to set the priority as a potential oddity. But I don't see what we can do about it - there is no place to set the priority now it is in the not-yet-existing FieldHolder. Also any such change of priority is either racing with a successful completion of the thread constructor, which will also set priority, or racing with a failed attach in which case the priority value is irrelevant. 

It is a general flaw that JVMTI allows you to get a hold of a partially initialized thread, and race with the initialization of that thread (the spec says nothing about this possibility - nor should it). So trying to reason what is correct behaviour in such circumstances is very hard.

> it suggests there the error handling might isn't complete somewhere

I don't understand what you mean by this.

**Comment by dholmes-ora** [comment]
Actually if the JVMTI code invokes Thread.setPriority then it will either throw NPE because holder is still null, or else work fine because holder has now been set.

**Comment by dholmes-ora** [comment]
I think I have over-thought this. Now that `allocate_threadObj` is fixed I don't think any of the setter functions can ever actually be called on an attaching thread. I'll double-check in the morning,

**Comment by dholmes-ora** [comment]
I re-examined the setter methods and none of them can be called on a JNI attaching thread whilst it is partially constructed. Such a thread could have regular Java methods called on it, but they would throw NPE when trying to access the FieldHolder. So it suffices to keep the assertion that holder!=nullptr in the setter methods.

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 1]
```diff
@@ -1630,25 +1632,42 @@ void java_lang_Thread::set_name(oop java_thread, oop name) {
   java_thread->obj_field_put(_name_offset, name);
 }
 
+// Convenience macros for setting and getting Thread fields that
+// are actually stored in the FieldHolder object of the thread.
+// The FieldHolder can be null whilst a thread is attaching via JNI.
```
Comment (AlanBateman) [inline src/hotspot/share/classfile/javaClasses.cpp:1637]: Also the primordial thread. A JVMTI agent with the can_generate_early_vmstart capability sees the start phase before create_initial_thread and so many instrument some classes that are loaded early in the startup. Agents playing in this phase of startup are playing with fire of course.

**Review by AlanBateman — APPROVED** [review]

**Comment by dholmes-ora** [comment]
Thanks for the review @AlanBateman !

**Review by rose00 — COMMENTED** [review]
Yes, that looks reasonable.  The macro regularizes the necessary careful treatment of the extra indirection.  I see how a previous one-off attempt probably left holes, which this systematic treatment would fill.

Changing CHECK to THREAD is tolerable, since not much happens afterwards, just a statement or two.  Maybe add a comment saying what could fail, and why that extra logic is still OK to run even if the failure happens.

**Comment by dholmes-ora** [comment]
Thanks for taking a look at this @rose00 .

> Changing CHECK to THREAD is tolerable, since not much happens afterwards, just a statement or two. Maybe add a comment saying what could fail, and why that extra logic is still OK to run even if the failure happens.

I've changed THREAD to CHECK so that we will skip the following statements. The call to `java_lang_Thread::set_daemon(thread_oop())` could assert due to the null FieldHolder.

**Review by AlanBateman — APPROVED** [review]

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 2]
```diff
@@ -1599,7 +1600,8 @@ void java_lang_Thread::clear_scopedValueBindings(oop java_thread) {
 }
 
 oop java_lang_Thread::holder(oop java_thread) {
-    return java_thread->obj_field(_holder_offset);
+  // Note: may be null if the thread is still attaching
```
Comment (dcubed-ojdk) [inline src/hotspot/share/classfile/javaClasses.cpp:1603]: nit perhaps: s/may be null/may return null/

File: src/hotspot/share/classfile/javaClasses.cpp [hunk 3]
```diff
@@ -1700,12 +1712,7 @@ JavaThreadStatus java_lang_Thread::get_thread_status(oop java_thread) {
   assert(Threads_lock->owned_by_self() || Thread::current()->is_VM_thread() ||
          JavaThread::current()->thread_state() == _thread_in_vm,
          "Java Thread is not running in vm");
-  oop holder = java_lang_Thread::holder(java_thread);
-  if (holder == nullptr) {
-    return JavaThreadStatus::NEW;  // Java Thread not initialized
-  } else {
-    return java_lang_Thread_FieldHolder::get_thread_status(holder);
-  }
+  GET_FIELDHOLDER_FIELD(java_thread, get_thread_status, JavaThreadStatus::NEW;  /* not initialized */);
```
Comment (dcubed-ojdk) [inline src/hotspot/share/classfile/javaClasses.cpp:1715]: I didn't expect to see a ';' after the 'NEW'. Should that be deleted?
Should it have caused a compile error?

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up. My only comments are nits.

I like how much cleaner the code is now.

File: src/hotspot/share/classfile/javaClasses.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/javaClasses.cpp]: Changed

File: src/hotspot/share/classfile/javaClasses.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/javaClasses.cpp]: Well spotted. It isn't wanted but doesn't cause an error - not sure why not as it is an unreachable empty statement. Anyway it is fixed.

**Comment by dholmes-ora** [comment]
That for the review @dcubed-ojdk !

/integrate

**Pull request closed without merging**